belt-dwp/src/lib.rs

```diff
@@ -0,0 +1,274 @@
+//! cipher.decrypt_in_place(&nonce, b"", &mut buffer)?;
+//! assert_eq!(&buffer, b"plaintext message");
+//! # Ok(())
+//! # }
+//! ```
+//!
+//! Similarly, enabling the `arrayvec` feature of this crate will provide an impl of
+//! [`aead::Buffer`] for `arrayvec::ArrayVec` (re-exported from the [`aead`] crate as
+//! [`aead::arrayvec::ArrayVec`]).
+
+use aead::consts::{U8, U16, U32};
+pub use aead::{self, AeadCore, AeadInPlace, Error, Key, KeyInit, KeySizeUser};
+use aead::{AeadInPlaceDetached, PostfixTagged};
+use belt_block::cipher::{Block, BlockCipherEncrypt, KeyIvInit, StreamCipher};
+use belt_block::{BeltBlock, belt_block_raw};
+use belt_ctr::BeltCtr;
+use universal_hash::UniversalHash;
+
+use crate::{
+    ghash::GHash,
+    utils::{from_u32, to_u32},
+};
+
+/// Nonce type for [`BeltDwp`]
+pub type Nonce = aead::Nonce<BeltDwp>;
+
+/// Tag type for [`BeltDwp`]
+pub type Tag = aead::Tag<BeltDwp>;
+
+mod gf;
+mod ghash;
+mod utils;
+
+/// T from the STB 34.101.31-2020
+const T: u128 = 0xE45D_4A58_8E00_6D36_3BF5_080A_C8BA_94B1;
+
+/// Belt-DWP authenticated encryption with associated data (AEAD) cipher, defined in
+/// STB 34.101.31-2020
+pub struct BeltDwp {
+    key: Key<BeltBlock>,
+}
+
+impl KeySizeUser for BeltDwp {
+    type KeySize = U32;
+}
+
+impl AeadInPlaceDetached for BeltDwp {
+    fn encrypt_in_place_detached(
+        &self,
+        nonce: &Nonce,
+        associated_data: &[u8],
+        buffer: &mut [u8],
+    ) -> aead::Result<Tag> {
+        Cipher::new(self.key, nonce).encrypt_in_place_detached(associated_data, buffer)
+    }
+
+    fn decrypt_in_place_detached(
+        &self,
+        nonce: &Nonce,
+        associated_data: &[u8],
+        buffer: &mut [u8],
+        tag: &Tag,
+    ) -> aead::Result<()> {
+        Cipher::new(self.key, nonce).decrypt_in_place_detached(associated_data, buffer, tag)
+    }
+}
+
+impl PostfixTagged for BeltDwp {}
+
+struct Cipher {
+    enc_cipher: BeltCtr,
+    mac_cipher: BeltBlock,
+    ghash: GHash,
+}
+
+impl Cipher {
+    fn new(key: Key<BeltBlock>, nonce: &Nonce) -> Self {
+        let cipher: BeltCtr = BeltCtr::new(&key, nonce);
+
+        let _s = to_u32::<4>(nonce);
+        let _k = to_u32::<8>(&key);
+        // 2.1. 𝑠 ← belt-block(𝑆, 𝐾);
+        let s = belt_block_raw(_s, &_k);
+        // 2.2. 𝑟 ← belt-block(𝑠, 𝐾);
+        let r = from_u32::<16>(&belt_block_raw(s, &_k));
+
+        Self {
+            enc_cipher: cipher,
+            mac_cipher: BeltBlock::new(&key),
+            // Unwrap is safe because the key is always 16 bytes
+            ghash: GHash::new_with_init_block(&Key::<GHash>::try_from(&r[..]).unwrap(), T),
+        }
+    }
+
+    fn encrypt_in_place_detached(
+        &mut self,
+        associated_data: &[u8],
+        buffer: &mut [u8],
+    ) -> aead::Result<Tag> {
+        let sizes_block =
+            get_sizes_block(associated_data.len() as u64 * 8, buffer.len() as u64 * 8);
```

Comment:
Introduce separate variables to make it a bit more readable?
```rust
let aad_bit_len = associated_data.len() as u64 * 8;
let msg_bit_len = buffer.len() as u64 * 8;
```

Reply:
Sure. Fixed.